File: MemoryGame/Models/GameEnums.swift
//
//  GameEnums.swift
//  Memory Match Kids
//

import Foundation
import SwiftUI

enum GridSize: String, CaseIterable, Identifiable {
    case twoByTwo = "2×2"
    case twoByThree = "2×3"
    case twoByFour = "2×4"
    case threeByFour = "3×4"
    case fourByFour = "4×4"
    case fourByFive = "4×5"
    case fiveBySix = "5×6"

    var id: String { rawValue }

    var gridRows: Int {
        switch self {
        case .twoByTwo: return 2
        case .twoByThree: return 2
        case .twoByFour: return 2
        case .threeByFour: return 3
        case .fourByFour: return 4
        case .fourByFive: return 4
        case .fiveBySix: return 5
        }
    }

    var gridColumns: Int {
        switch self {
        case .twoByTwo: return 2
        case .twoByThree: return 3
        case .twoByFour: return 4
        case .threeByFour: return 4
        case .fourByFour: return 4
        case .fourByFive: return 5
        case .fiveBySix: return 6
        }
    }

    var pairCount: Int { (gridRows * gridColumns) / 2 }
}

enum MatchMode: String, Codable {
    case identical
    case association
    case triple
    case sequence
    case maze
}

enum BadgeTier: String, CaseIterable, Codable {
    case bronze = "Bronze"
    case silver = "Silver"
    case gold = "Gold"

    var icon: String {
        switch self {
        case .bronze: return "medal.fill"
        case .silver: return "medal.fill"
        case .gold: return "crown.fill"
        }
    }

    var colorHex: String {
        switch self {
        case .bronze: return "CD7F32"
        case .silver: return "C0C0C0"
        case .gold: return "FFD700"
        }
    }

    static func forStars(_ stars: Int) -> BadgeTier? {
        if stars >= 3 { return .gold }
        if stars >= 2 { return .silver }
        if stars >= 1 { return .bronze }
        return nil
    }
}

enum AppearanceMode: String, CaseIterable, Identifiable, Codable {
    case system
    case light
    case dark

    var id: String { rawValue }

    var title: String {
        switch self {
        case .system: return "System"
        case .light: return "Light"
        case .dark: return "Dark"
        }
    }

    var icon: String {
        switch self {
        case .system: return "circle.lefthalf.filled"
        case .light: return "sun.max.fill"
        case .dark: return "moon.fill"
        }
    }

    var preferredColorScheme: ColorScheme? {
        switch self {
        case .system: return nil
        case .light: return .light
        case .dark: return .dark
        }
    }
}

File: android/app/src/main/java/com/memogame/app/model/GameEnums.kt
package com.memogame.app.model

enum class GridSize(val label: String, val gridRows: Int, val gridColumns: Int) {
    TWO_BY_TWO("2×2", 2, 2),
    TWO_BY_THREE("2×3", 2, 3),
    TWO_BY_FOUR("2×4", 2, 4),
    THREE_BY_FOUR("3×4", 3, 4),
    FOUR_BY_FOUR("4×4", 4, 4),
    FOUR_BY_FIVE("4×5", 4, 5),
    FIVE_BY_SIX("5×6", 5, 6);

    val pairCount: Int get() = (gridRows * gridColumns) / 2
}

enum class MatchMode {
    IDENTICAL,
    ASSOCIATION,
    TRIPLE
}

enum class BadgeTier(val rawValue: String, val colorHex: String) {
    BRONZE("Bronze", "CD7F32"),
    SILVER("Silver", "C0C0C0"),
    GOLD("Gold", "FFD700");

    companion object {
        fun forStars(stars: Int): BadgeTier? = when {
            stars >= 3 -> GOLD
            stars >= 2 -> SILVER
            stars >= 1 -> BRONZE
            else -> null
        }

        fun fromRaw(raw: String?): BadgeTier? = entries.firstOrNull { it.rawValue == raw }
    }
}

enum class AppearanceMode(val rawValue: String, val title: String) {
    SYSTEM("system", "System"),
    LIGHT("light", "Light"),
    DARK("dark", "Dark");

    companion object {
        fun fromRaw(raw: String?): AppearanceMode =
            entries.firstOrNull { it.rawValue == raw } ?: SYSTEM
    }
}

enum class CardBackStyle(val rawValue: String, val label: String, val colorHexes: List<String>) {
    CLASSIC("classic", "Classic", listOf("5B8DEF", "7B5BEF", "9B4DEF")),
    OCEAN("ocean", "Ocean", listOf("2BC0E4", "1A6FB5")),
    SUNSET("sunset", "Sunset", listOf("FF9500", "FF5E3A", "FF2D55")),
    FOREST("forest", "Forest", listOf("34C759", "1A8F50")),
    CANDY("candy", "Candy", listOf("FF6B9D", "C44DFF"));

    companion object {
        fun fromRaw(raw: String?): CardBackStyle =
            entries.firstOrNull { it.rawValue == raw } ?: CLASSIC
    }
}
